lib/models/os.coffee

```diff
@@ -14,23 +14,223 @@ See the License for the specific language governing permissions and
 limitations under the License.
 ###
 
+Promise = require('bluebird')
+reject = require('lodash/reject')
+once = require('lodash/once')
+semver = require('semver')
+
 errors = require('resin-errors')
-{ onlyIf } = require('../util')
+
+{ onlyIf, getImgMakerHelper, findCallback, notFoundResponse, deviceTypes: deviceTypesUtil } = require('../util')
 
 getOsModel = (deps, opts) ->
 	{ request } = deps
 	{ isBrowser, imageMakerUrl } = opts
 
+	imgMakerHelper = getImgMakerHelper(imageMakerUrl, request)
+
+	configModel = once -> require('./config')(deps, opts)
+
+	getDeviceTypes = once ->
+		configModel().getDeviceTypes()
+
+	isValidDeviceType = (deviceType) ->
+		getDeviceTypes()
+		.then (types) ->
+			!!deviceTypesUtil.findBySlug(types, deviceType)
+
+	getDownloadSize = imgMakerHelper.buildApiRequester
+		url: '/size_estimate'
+		withVersion: true
+		postProcess: ({ body }) -> body.size
+
+	getOsVersions = imgMakerHelper.buildApiRequester
+		buildUrl: ({ deviceType }) ->
+			"/image/#{deviceType}/versions"
+		postProcess: ({ body }) ->
+			{ versions, latest } = body
```

Comment:
Nit: you could extract these directly above: `({ body: { versions, latest } }) ->`